numpy/linalg/_linalg.py

```diff
@@ -2068,18 +2068,23 @@ def matrix_rank(A, tol=None, hermitian=False, *, rtol=None):
     >>> matrix_rank(np.zeros((4,)))
     0
     """
+    if rtol is not None and tol is not None:
+        raise ValueError("`tol` and `rtol` can't be both set.")
+
     A = asarray(A)
     if A.ndim < 2:
         return int(not all(A == 0))
     S = svd(A, compute_uv=False, hermitian=hermitian)
-    if rtol is not None and tol is not None:
-        raise ValueError("`tol` and `rtol` can't be both set.")
+
     if rtol is None:
```

Comment:
Could you still move this and the next two lines inside the `if tol is None` branch? Also, since in this branch, `rtol` is guaranteed to be a scalar, it can be,
```
if tol is None:
    if rtol is None:
        rtol = max(A.shape[-2:]) * finfo(S.dtype).eps
    else:
        rtol = asarray(rtol)[..., newaxis]
    tol = S.max(axis=-1, keepdims=True) * rtol
else:
    ...
```

Reply:
Ah right, I confused `tol` with `rtol` in your previous comment - now I should be right.